### ISSUE TITLE
Some issue with exit nodes

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#4010]
In https://github.com/tailscale/tailscale/issues/2654#issuecomment-1048431262, @karan wrote:

> Apologies for the late response but I tried unstable and it did not fix the issue for me:

```
$ curl -vvv ipinfo.io
*   Trying 34.117.59.81:80...
^C
```

> Even stuff on LAN doesn't work (MBP has allow LAN enabled).
>
> Bug report from 1.21.120 exit node: `BUG-5b731bbefdc71d1ad493765d666427d604a949402a0938b0e1ae63ac60fb015f-20220223040412Z-9b42bd72c49035a2`
>
> Bug report from 1.20.2 MBP: `BUG-5f03ad43e99c98ced11182aa2ea3fc013c9752b71e46c34aa3cd2efe42f031a8-20220223040450Z-83dd3d4f9ccbd24c`

---

I'm moving this bug here because #2654 got out of control and is also seemingly fixed for most people, so I'll be closing that since we identified the main problem that most people were hitting.

**Comment by bradfitz** [comment]
@karan, we're in the same time zone (actually both in Seattle), so perhaps we can just debug this over a call sometime this week?

Looks like you're running the exit node in Docker? And you've disabled IPv6 explicitly? Those should both be fine. Just looking.

Oh, and you're using `--accept-dns=false`. That should also be fine, but some combination of the three (along with being an exit node) might be hitting some edge case.

From your Mac, can you `curl http://100.85.179.103:48329/dns-query?q=google.com` ?

(Aside: why did you disable IPv6 to begin with?)

**Comment by karan** [comment]
Thanks Brad!

> @karan, we're in the same time zone (actually both in Seattle), so perhaps we can just debug this over a call sometime this week?

Be happy to if needed.

> Looks like you're running the exit node in Docker? And you've disabled IPv6 explicitly? Those should both be fine. Just looking.

Exit node in question (`100.85.179.103`) is not in Docker, it's just a systemd unit on headless ubuntu. A different node `100.78.12.32` is in Docker but that's not the one I'm using.

> Oh, and you're using `--accept-dns=false`. That should also be fine, but some combination of the three (along with being an exit node) might be hitting some edge case.

I think `--accept-dns=false` was added because I disabled MagicDNS as one of my debugging<sup>trial and error</sup> steps.

> From your Mac, can you `curl http://100.85.179.103:48329/dns-query?q=google.com` ?

Sure. With and without exit node being used, I get the same result:

```
$ curl http://100.85.179.103:48329/dns-query?q=google.com
["142.250.217.110"]
```

What service is running at 48329?

> (Aside: why did you disable IPv6 to begin with?)

No particular reason. I don't remember explicitly disabling it though. Why and how should I enable it?

One more data point: Interestingly, my Android node running 1.20.2 is able to work perfectly fine using the same exit node (`.103`). So this _may_ be a Mac issue (App Store lists 1.20.4 as the latest version but there's no update button).

**Comment by bradfitz** [comment]
> What service is running at 48329?

It's a random port. That's just what yours currently happens to be. That's the Tailscale "peer API" that's used between Tailscale nodes for misc stuff like Taildrop and being a DNS proxy when exit nodes are in use.

From your Mac when exit node is on, can you curl 1.1.1.1 or ping 8.8.8.8?

Does your Mac have "Use Tailscale DNS" checked? Forget whether that does anything with exit nodes but worth trying both ways.

**Comment by karan** [comment]
Okay here's some stuff:

```
# Exit node ON, Use Tailscale DNS ON
$ curl 1.1.1.1
^C

$ ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8): 56 data bytes
64 bytes from 8.8.8.8: icmp_seq=0 ttl=116 time=10.591 ms
64 bytes from 8.8.8.8: icmp_seq=1 ttl=116 time=14.272 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=116 time=11.077 ms
64 bytes from 8.8.8.8: icmp_seq=3 ttl=116 time=11.735 ms
64 bytes from 8.8.8.8: icmp_seq=4 ttl=116 time=10.596 ms
64 bytes from 8.8.8.8: icmp_seq=5 ttl=116 time=11.676 ms
64 bytes from 8.8.8.8: icmp_seq=6 ttl=116 time=13.126 ms
64 bytes from 8.8.8.8: icmp_seq=7 ttl=116 time=11.519 ms
64 bytes from 8.8.8.8: icmp_seq=8 ttl=116 time=14.918 ms
^C
--- 8.8.8.8 ping statistics ---
9 packets transmitted, 9 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 10.591/12.168/14.918/1.487 ms

$ dig google.com
^C

# Exit node ON, Use Tailscale DNS OFF
$ curl 1.1.1.1
^C

$ ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8): 56 data bytes
64 bytes from 8.8.8.8: icmp_seq=0 ttl=116 time=13.255 ms
64 bytes from 8.8.8.8: icmp_seq=1 ttl=116 time=11.383 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=116 time=13.008 ms
^C
--- 8.8.8.8 ping statistics ---
3 packets transmitted, 3 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 11.383/12.549/13.255/0.830 ms

$ dig google.com

; <<>> DiG 9.10.6 <<>> google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 57073
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;google.com.			IN	A

;; ANSWER SECTION:
google.com.		293	IN	A	142.250.217.110

;; Query time: 84 msec
;; SERVER: 100.100.100.100#53(100.100.100.100)
;; WHEN: Tue Feb 22 21:12:47 PST 2022
;; MSG SIZE  rcvd: 55
```

**Comment by karan** [comment]
Eh so I tried exit node with my Android phone again and it doesn't work. 

Bug report BUG-9b98700c8703107c92fa60b3a8c4a0c8daec0bf952918236a8f56a35ab17f193-20220223192621Z-aa72278125666725

**Comment by FedX-sudo** [comment]
Upgraded a server and my local machine to 1.22, and exit nodes are not working. I have tried all the suggestions in the previous thread. A potentially interesting thing to note is that I am running PiHole over TailScale following the official guide. I am also overriding local DNS. However, it does not appear to be a DNS issue because it resolves domains just fine, but never get's a response. Thanks for all the work, and I don't know what I would do without TailScale!

Bug report BUG-9e3607d0e837ae71f085049f3b0b63c285e5e78063db8974568dade2334985b0-20220319220257Z-0388c73028b82f91

**Comment by mcfedr** [comment]
I seem to be having a similar issue, trying to use my linux server as an exit node from my mac.

with the exit node enabled


I still have DNS working on the mac, and can see using a tailscale dns resolver
```
$ dig google.com

; <<>> DiG 9.10.6 <<>> google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 63324
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4095
;; QUESTION SECTION:
;google.com.			IN	A

;; ANSWER SECTION:
google.com.		222	IN	A	216.58.208.206

;; Query time: 59 msec
;; SERVER: 100.100.100.100#53(100.100.100.100)
;; WHEN: Tue Aug 30 22:50:05 BST 2022
;; MSG SIZE  rcvd: 55
```

But cannot connect to internet, i.e. curl just hangs:

```
curl google.com
^C
```

by ip the same, just hangs

```
curl 216.58.208.206
^C
```

traceroute shows traffic going to the exit node

```
traceroute to google.com (216.58.208.206), 64 hops max, 52 byte packets
 1  gandi (100.121.121.127)  54.507 ms  53.490 ms  54.135 ms
^C
```

I made a bug report from my mac, 
BUG-2df16fc8fdb5382122ee96da26c7c81b271abe0f6244be7a6d871c679128770e-20220830215322Z-6bac3b2c5930086f

and from my linux server
BUG-4d9897018902d79e7f7926e364623a1d0a772d5f256aa3e45d6a7e76c861726d-20220830215345Z-01ca0fb783658b40

**Comment by bradfitz** [comment]
There was a good exit node fix in 1.30.1. I'm going to close this for now, as it's old and has a mix of issues in it.

Please open new bugs with details if anybody is still experiencing exit node issues with 1.30.1+.
